### PR TITLE
fix(oauth2): remove response_mode from authorization request v1.3

### PR DESCRIPTION
**Pull request by flaviodelgrosso**
This pull request introduces comprehensive updates to the OAuth2 authorization request test suite and core logic, primarily to support and validate differences between v1.0 and v1.3 of the ItWallet specification. The changes ensure correct handling of the `response_mode` parameter: it is required for v1.0 requests but must be omitted for v1.3 requests. The updates include stricter type checks, new and updated tests, and adjustments to core request-building logic.

Resolves WLEO-1156

Key changes include:

**Support for v1.3 specification and response_mode handling:**

* Updated tests and core logic to ensure that `response_mode` is not included in v1.3 authorization requests, with new tests verifying its absence and type-level checks to reject its usage in v1.3 options. [[1]](diffhunk://#diff-d6142b6b0eab3ce21ff2df1632964021a9fff1d6a5efffc1f3ca67b54a99cdacR1-R4) [[2]](diffhunk://#diff-d6142b6b0eab3ce21ff2df1632964021a9fff1d6a5efffc1f3ca67b54a99cdacR235-R267) [[3]](diffhunk://#diff-d6142b6b0eab3ce21ff2df1632964021a9fff1d6a5efffc1f3ca67b54a99cdacR417-R439) [[4]](diffhunk://#diff-8493c359c9aea37967f2f642146bed82f1e298718782c3daf3710bc68851e322R369-R422) [[5]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR77-R131) [[6]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR191-R208)
* Added configuration objects (`IoWalletSdkConfig`) for both v1.0 and v1.3 in test setups, and passed them explicitly to relevant functions to drive version-specific behavior. [[1]](diffhunk://#diff-d6142b6b0eab3ce21ff2df1632964021a9fff1d6a5efffc1f3ca67b54a99cdacR42-R49) [[2]](diffhunk://#diff-d6142b6b0eab3ce21ff2df1632964021a9fff1d6a5efffc1f3ca67b54a99cdacR64) [[3]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR15-R20) [[4]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR53) [[5]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR141) [[6]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR162) [[7]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR182) [[8]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR227) [[9]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR254) [[10]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR272) [[11]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR303) [[12]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR324) [[13]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR344) [[14]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR361) [[15]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR397) [[16]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR420)

**Type safety and stricter validation:**

* Refactored `CreatePushedAuthorizationRequestOptions` to be versioned and generic, enforcing at compile time that `responseMode` is not allowed for v1.3 requests. [[1]](diffhunk://#diff-095deb79e1a2e821a363a87824a9460e6e344ad4325c338bcda48d4bb61775e8R17-R32) [[2]](diffhunk://#diff-095deb79e1a2e821a363a87824a9460e6e344ad4325c338bcda48d4bb61775e8R67-R68)
* Improved validation in `parsePushedAuthorizationRequest` to throw errors for missing `response_mode` in v1.0, and to accept its absence in v1.3, with corresponding test coverage.

**Test coverage improvements:**

* Added/updated tests to ensure correct parsing and handling of unsigned and signed requests for both v1.0 and v1.3, including tests for error cases and edge conditions. [[1]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR77-R131) [[2]](diffhunk://#diff-916b8e327251f16ea41959c021984e6cf6f002917c5028c3792c9e90527cd79aR191-R208) [[3]](diffhunk://#diff-8493c359c9aea37967f2f642146bed82f1e298718782c3daf3710bc68851e322R369-R422)

**Internal code structure and imports:**

* Updated imports and internal type usage to support versioned request types and schemas, improving maintainability and clarity. [[1]](diffhunk://#diff-095deb79e1a2e821a363a87824a9460e6e344ad4325c338bcda48d4bb61775e8R5-R7) [[2]](diffhunk://#diff-095deb79e1a2e821a363a87824a9460e6e344ad4325c338bcda48d4bb61775e8R17-R32)

These changes ensure the codebase is robust against version-specific requirements and regressions, and that future updates to the ItWallet specification can be incorporated with minimal risk.

### CODE REVIEW

File: packages/oauth2/src/authorization-request/create-authorization-request.ts [hunk 1]
```diff
@@ -135,7 +171,8 @@ export interface CreatePushedAuthorizationRequestOptions {
  * @param options.jti - Optional JWT ID for PAR (auto-generated if not provided)
  * @param options.pkceCodeVerifier - Optional PKCE code verifier (auto-generated if not provided)
  * @param options.redirectUri - Redirect URI for the authorization response
- * @param options.responseMode - Response mode (must be supported by Credential Issuer)
+ * @param options.config - Italian Wallet specification version used to build the request shape
+ * @param options.responseMode - Response mode (v1.0 only, must be supported by Credential Issuer)
  * @param options.scope - OAuth 2.0 scope to request
```
[the remark this answers is not in the document]
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/create-authorization-request.ts:176]: [7d690b0](https://github.com/pagopa/io-wallet-sdk/pull/140/commits/7d690b01e20bed178a888d39ed76beba61876964)

File: packages/oauth2/src/authorization-request/parse-pushed-authorization-request.ts [hunk 2]
```diff
@@ -20,11 +23,14 @@ import {
 import {
   AuthorizationRequest,
   zAuthorizationRequest,
```
[the remark this answers is not in the document]
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/parse-pushed-authorization-request.ts]: Since this repo uses Zod 4, `z.SafeParseReturnType` is not exported, so I used `z.ZodSafeParseResult<AuthorizationRequest>` instead. [163348a](https://github.com/pagopa/io-wallet-sdk/pull/140/commits/163348a162d5cb4b2850c068a7fc2c6f9d5c3292)

File: packages/oauth2/src/authorization-request/z-authorization-request.ts [hunk 3]
```diff
@@ -43,6 +55,20 @@ export const zAuthorizationRequest = z
...
+
+export const zAuthorizationRequestV1_3 = zAuthorizationRequestBase;
+export type AuthorizationRequestV1_3 = z.infer<
+  typeof zAuthorizationRequestV1_3
+>;
+
+export const zAuthorizationRequest = z.union([
+  zAuthorizationRequestV1_0,
+  zAuthorizationRequestV1_3,
+]);
```
[the remark this answers is not in the document]
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/z-authorization-request.ts]: The important fix is that response_mode will be required only for v1.0. It's fine to not have a strictly object shape

**Comment by flaviodelgrosso** [comment]
@manuraf When `parsePushedAuthorizationRequest` runs with a v1.3 config, it uses this new schema. However, because the shared base is a `looseObject`, a v1.3 PAR or JAR payload that still includes `response_mode` is accepted, and the field is preserved.

The spec does not explicitly say that `response_mode` must not be present. Still, if we want to enforce that, we would need to refine the schema for this field specifically.

My assumption is that, if we are intentionally using `looseObject`, we do not need to enforce that restriction, and instead should only make sure that `response_mode` is required in v1.0.

File: .changeset/curly-carrots-draw.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@pagopa/io-wallet-oauth2": patch
```
Comment (manuraf) [inline .changeset/curly-carrots-draw.md]: The changeset currently declares this as a `patch` release, but the PR adds new **required** fields to two public option types — `ParsePushedAuthorizationRequestOptions.config` and the reworked `CreatePushedAuthorizationRequestOptions` discriminated union (where `responseMode` is now v1.0-only and `config` is required on both variants). Existing consumers relying on the previous flat option shape will fail at compile time, which is a textbook breaking change regardless of the intended scope.

Looking at prior PRs in this repo (#137, #136), the convention has been to treat required-field additions as `minor` on the `1.x` line. You could follow that same precedent here and bump this entry to `minor`, updating the description to explicitly call out the two affected types so downstream consumers know what changed:

```md
"@pagopa/io-wallet-oauth2": minor
```

If the team decides to be strictly semver-correct, `major` is the other option — but `patch` understates the impact for anyone on a `^1.3.0` range.

File: packages/oauth2/src/authorization-request/z-authorization-request.ts [hunk 5]
```diff
@@ -43,6 +55,20 @@ export const zAuthorizationRequest = z
...
+export type AuthorizationRequestV1_0 = z.infer<
+  typeof zAuthorizationRequestV1_0
+>;
+
+export const zAuthorizationRequestV1_3 = zAuthorizationRequestBase;
+export type AuthorizationRequestV1_3 = z.infer<
+  typeof zAuthorizationRequestV1_3
+>;
+
+export const zAuthorizationRequest = z.union([
```
Comment (manuraf) [inline packages/oauth2/src/authorization-request/z-authorization-request.ts]: The `zAuthorizationRequest` union here allows a v1.0 input that omits `response_mode` to silently pass validation via the v1.3 branch — bypassing the enforcement that v1.0 is supposed to require. This was [empirically verified](https://github.com/pagopa/io-wallet-sdk/pull/140) and the same widening affects `zPushedAuthorizationRequestUnsigned` / `PushedAuthorizationRequestUnsigned` at line 130–134.

Since no internal caller depends on the union semantics anymore (all have been switched to version-specific schemas in this PR), you could remove `zAuthorizationRequest`, `AuthorizationRequest`, `zPushedAuthorizationRequestUnsigned`, and `PushedAuthorizationRequestUnsigned` from the public API entirely, forcing consumers to choose a version-specific schema. This eliminates the silent-acceptance footgun and keeps the API surface consistent with the version-aware design introduced here.

File: packages/oauth2/src/authorization-request/create-authorization-request.ts [hunk 6]
```diff
@@ -285,12 +346,41 @@ export async function createPushedAuthorizationRequest(
...
   }
+
+  return {
+    authorizationRequest,
+    client_id: options.clientId,
+    pkceCodeVerifier: pkce.codeVerifier,
+  };
+}
+
+function parseAuthorizationRequestByVersion(
```
Comment (manuraf) [inline packages/oauth2/src/authorization-request/create-authorization-request.ts:359]: The `version` local in `parseAuthorizationRequestByVersion` is only ever read inside the `default` throw — it doesn't contribute to type narrowing since the switch already discriminates directly on `options.config.itWalletSpecsVersion`. You could drop the intermediate binding and inline the access directly in the `default` branch to keep things tighter and remove a binding that suggests broader reuse.

File: packages/oauth2/src/authorization-request/create-authorization-request.ts [hunk 7]
```diff
@@ -285,12 +346,41 @@ export async function createPushedAuthorizationRequest(
...
+  options: CreatePushedAuthorizationRequestOptions,
+  baseAuthorizationRequest: Omit<
+    AuthorizationRequest,
+    "authorization_details" | "response_mode" | "scope"
+  > &
+    Pick<AuthorizationRequest, "authorization_details" | "scope">,
+): AuthorizationRequest {
+  const version = options.config.itWalletSpecsVersion;
+
+  if (hasConfigVersion(options, ItWalletSpecsVersion.V1_0)) {
```
Comment (manuraf) [inline packages/oauth2/src/authorization-request/create-authorization-request.ts:369]: This file uses `hasConfigVersion(options, ...)` while the neighbouring `parse-pushed-authorization-request.ts` uses `options.config.isVersion(...)` — both are functionally identical, but mixing the two idioms within the same feature makes the codebase harder to scan.

You could standardize on `options.config.isVersion(...)` throughout the `authorization-request/` subtree, since it co-locates the version check with the config object and reads a bit more naturally at call sites:

```typescript
if (options.config.isVersion(ItWalletSpecsVersion.V1_0)) {
  // ...
}
if (options.config.isVersion(ItWalletSpecsVersion.V1_3)) {
  // ...
}
```

File: .changeset/curly-carrots-draw.md (hunk 4 above)
Comment (flaviodelgrosso) [reply .changeset/curly-carrots-draw.md]: [fffe45f](https://github.com/pagopa/io-wallet-sdk/pull/140/commits/fffe45f520c97549b35ffec2917f3bfbd3498b1b)

File: packages/oauth2/src/authorization-request/z-authorization-request.ts (hunk 5 above)
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/z-authorization-request.ts]: [449d3ed](https://github.com/pagopa/io-wallet-sdk/pull/140/commits/449d3ed31445ef6da45a3fe370f10ed0eccc8ae7)

File: packages/oauth2/src/authorization-request/create-authorization-request.ts (hunk 6 above)
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/create-authorization-request.ts:359]: It is true that version is only used in the final ItWalletSpecsVersionError, but dropping it and inlining `options.config.itWalletSpecsVersion` in the throw is likely not a safe tightening here. The two `hasConfigVersion(...)` checks are type guards. Since `CreatePushedAuthorizationRequestOptions` only includes V1_0 | V1_3, after both guarded branches return, TypeScript can narrow options to never. The precomputed version preserves the runtime value for the defensive error path before that narrowing happens

File: packages/oauth2/src/authorization-request/create-authorization-request.ts (hunk 7 above)
Comment (flaviodelgrosso) [reply packages/oauth2/src/authorization-request/create-authorization-request.ts:369]: We need to keep `hasConfigVersion` here. Unlike in `parse-pushed-authorization-request.ts`, this branch needs to narrow the whole options union so options.responseMode is available only for V1_0. `options.config.isVersion(...)` narrows the config object, but it does not provide the same guarantee for version-specific fields on the containing options object. So the two forms are equivalent at runtime, but not equivalent for type narrowing in this module

**Review by manuraf — APPROVED** [review]
